Implement dynamic sensors for tplink

rytilahti opened the pull request:
## This PR is part of the tplink rewrite series, to be merged into `tplink_rewrite` branch (see #111151).
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This adds a base entity class shared by all platforms & converts `sensor` platform to use upstream-provided definitions in addition to the existing ones.
This is a backwards compatible with the existing sensors (emeter) and switches (led).

![image](https://github.com/home-assistant/core/assets/3705853/f99a2a25-fa9b-4d71-9b58-4495c402e1e7)



TODO:
- [x] Child device specific data update coordinators
- [x] Removing legacy handling altogether
- [ ] Tests for backwards compat of legacy identifiers
- [x] ~~Add translations~~ (can be done in a subsequent PR, to unblock other platforms for now)
- [x] Mapping for `(feature_id, device_class)` & co

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

rytilahti committed:
Implement dynamic sensors for tplink

rytilahti committed:
Fix some linting issues

sdb9696 committed:
Bump python-kasa to 0.7.0.dev5 (#119828)

Bump tplink kasa dependency to 0.7.0.dev5

rytilahti committed:
Use runtime_data

rytilahti committed:
todo--

sdb9696 committed:
Remove old sensor class, add tests, polish (#119499)

* Remove old sensor class and fix tests

* Use static descriptions if available

* Move checking of consumption_today is none into library

* Fix tests

* Add tests

* Re-consolidate sensor and switch entity creation logic

* Fix some docstring issues

* Support new datetime sensor

* Add descriptions module

* Replace static entity descriptions with extras

* Add readme to tplink integration

* Remove redundant light entity description

* Pull datetime sensor timezone handling

bdraco commented on homeassistant/components/tplink/entity.py:
```diff
@@ -181,24 +173,10 @@ def _get_unique_id(self) -> str:
 
         # For legacy sensors, we construct our IDs from the entity description
-        if self.entity_description is not None:
+        if isinstance(self, SensorEntity) and self.entity_description is not None:
```
the base entity should never know about the details of a platform

sdb9696 replied:
Done in https://github.com/home-assistant/core/pull/120025

bdraco commented on homeassistant/components/tplink/entity.py:
```diff
@@ -153,12 +146,11 @@ def __init__(
         """Return unique ID for the entity."""
         device = self._device
-        if self._feature is not None:
+        if not isinstance(self, SensorEntity) and self._feature is not None:
```
the base entity should never know about the details of a platform

bdraco replied:
One way around this is to overload `_get_unique_id` in the subclass based on the platform

sdb9696 replied:
Done in https://github.com/home-assistant/core/pull/120025

bdraco commented on homeassistant/components/tplink/entity.py:
```diff
@@ -28,12 +28,11 @@
 from homeassistant.helpers.update_coordinator import CoordinatorEntity
 
 from . import legacy_device_id
-from .const import DOMAIN, PRIMARY_STATE_ID
```
Any time we have an import of platform into the base entity module this is a hint that we have a design issue where the abstraction of the sensor and light platform is leaking into the base entity.
```
from homeassistant.components.light import LightEntity
from homeassistant.components.sensor import SensorEntity
```

sdb9696 replied:
Done in https://github.com/home-assistant/core/pull/120025

bdraco commented on homeassistant/components/tplink/entity.py:
```diff
@@ -274,55 +255,92 @@ def _entities_for_device_and_its_children[_E: CoordinatedTPLinkEntity](
+        kwargs["translation_key"] = feature.id
+        kwargs["state_class"] = entity_extras.get("state_class")
+        kwargs["device_class"] = entity_extras.get("device_class")
+    else:
+        kwargs["name"] = feature.name
+        kwargs["icon"] = feature.icon
```
How will this work with entity translations?

sdb9696 replied:
The entities that do not have descriptions setup will not be able to use translation and we will log a warning in https://github.com/home-assistant/core/pull/120025

rytilahti replied:
I noticed that if the translations are not generated when starting the homeassistant, we will get devices named `None`, but that should never concern regular users.

bdraco commented on homeassistant/components/tplink/sensor.py:
```diff
@@ -1,182 +1,79 @@
+            TPLinkSensorEntityDescription, feature
+        )
+        super().__init__(device, coordinator, feature=feature, parent=parent)
+        self._feature: Feature
```
```suggestion
```

bdraco commented on homeassistant/components/tplink/sensor.py:
```diff
@@ -1,182 +1,79 @@
+class TPLinkSensor(CoordinatedTPLinkEntity, SensorEntity):
+    """Representation of a feature-based TPLink sensor."""
 
     entity_description: TPLinkSensorEntityDescription
```
```suggestion
    entity_description: TPLinkSensorEntityDescription
    _feature: Feature
```

sdb9696 replied:
Done in https://github.com/home-assistant/core/pull/120025

bdraco commented on homeassistant/components/tplink/sensor.py:
```diff
@@ -1,182 +1,79 @@
 
-class SmartPlugSensor(CoordinatedTPLinkEntity, SensorEntity):
-    """Representation of a TPLink Smart Plug energy sensor."""
+class TPLinkSensor(CoordinatedTPLinkEntity, SensorEntity):
```
Would be cleaner to subclass `CoordinatedTPLinkEntity` as `CoordinatedTPLinkFeatureEntity` so `self._feature` is typed in the base class to always exist

sdb9696 replied:
Done in https://github.com/home-assistant/core/pull/120025

bdraco commented on homeassistant/components/tplink/entity.py:
```diff
@@ -274,55 +255,92 @@ def _entities_for_device_and_its_children[_E: CoordinatedTPLinkEntity](
     )
+    kwargs["entity_category"] = _category_for_feature(feature)
+
+    if feature.type == Feature.Type.Sensor:
```
```suggestion
    if feature.type is Feature.Type.Sensor:
```

if both sides are an enum

bdraco commented on homeassistant/components/tplink/entity.py:
```diff
@@ -274,55 +255,92 @@ def _entities_for_device_and_its_children[_E: CoordinatedTPLinkEntity](
+    else:
+        kwargs["key"] = feature.id
+    # Use translation keys for all features configured in extras
+    if entity_extras:
```
Instead of extras, write out `EntityDescriptions` and use `dataclasses.replace` to create the dynamic atts

bdraco commented on homeassistant/components/tplink/entity.py:
```diff
@@ -274,55 +255,92 @@ def _entities_for_device_and_its_children[_E: CoordinatedTPLinkEntity](
-        **kwargs,
+    kwargs: dict = {}
+    # Key can be overridden for legacy support
+    if (entity_extras := ENTITY_EXTRAS.get(feature.id)) and (
```
Make this lookup a base EntityDescription and use `dataclasses.replace` on it below to add the additional keys

sdb9696 replied:
Done in https://github.com/home-assistant/core/pull/120025

bdraco commented on homeassistant/components/tplink/diagnostics.py:
```diff
@@ -42,10 +40,10 @@
 ) -> dict[str, Any]:
     """Return diagnostics for a config entry."""
-    data: TPLinkData = hass.data[DOMAIN][entry.entry_id]
+    data = entry.runtime_data
```
Minor: some places call this `entry` and others call it `config_entry`.  I prefer the shorter but we should pick one or the other and it doesn't matter so much as long as we are consistent

sdb9696 replied:
Done in https://github.com/home-assistant/core/pull/120025

rytilahti committed:
Add temperature & cleanup a bit

sdb9696 committed:
Use entity descriptions for feature overrides in tplink integration (#120025)

Remove entity extras from TPLink integration.
Remove any platform specific code from entity.py (which moves some of the unique id logic back to light but it makes it a lot more readable imo)

Addresses review comments in #117764
